feature_engine/variable_manipulation.py

```diff
@@ -127,6 +130,64 @@ def _find_or_check_categorical_variables(
     return variables
 
 
+def _find_or_check_datetime_variables(
+    X: pd.DataFrame, variables: Variables = None
+) -> List[Union[str, int]]:
+    """
+    Checks that variables provided by the user are of type datetime,
+    and transform date/times given in str/obj format into datetimes.
+    If None, finds all the datetime type variables in the DataFrame.
+
+    Parameters
+    ----------
+    X : pandas DataFrame
+    variables : variable or list of variables. Defaults to None
+
+    Returns
+    -------
+    variables : List of datetime variables
+    """
+
+    if not variables:
+        variables = [
+            column
+            for column in X.select_dtypes(exclude="number").columns
+            if is_datetime(X[column])
+            or (
+                is_object(X[column])
+                and is_datetime(pd.to_datetime(X[column], errors="ignore"))
+            )
+        ]
+
+        if len(variables) == 0:
+            raise ValueError(
+                "No datetime variables in this dataframe. "
+                "Note: purely numeric variables representing dates or times "
+                "will not be treated as datetime by this transformer."
+            )
+        return variables
+
+    if isinstance(variables, (str, int)):
```

Comment:
Both _find_or_check_numerical_variables and _find_or_check_categorical_variables actually [do](https://github.com/feature-engine/feature_engine/blob/main/feature_engine/variable_manipulation.py#L62) that check.
In fact, I think this along with the logic of the functions - if, elif, else structure - is a problem. When either of those _find_or_check* functions is passed a string or an int, it will
1. make a list that contains it
2. return that list
__without__ actually entering the checks! This has probably gone unnoticed because there are no tests where a string/int corresponding to a non-numerical or non-categorical variable is passed to its _find_or_check counterpart

Reply:
Actually you are right. We do need that there. 

I forgot I changed the logic of `_check_input_parameter_variables` when I added the checks for the check_estimators from sklearn.

Apologies :_(

Reply:
No problem, I'm still not too sure what to do though, I would be inclined to change the logic inside _find_or_check_numerical_variables and _find_or_check_categorical_variables aswell. Please have a look at the demo below (on the PowerTransformer, which I haven't touched in my fork) <br>
Here, _variables_ is passed as a list and the correct TypeError is thrown at the _find_or_check_numerical_variables level:
```
>>> pt = PowerTransformer(variables=["City"])
>>> pt.fit(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\feature_engine\feature_engine\transformation\power.py", line 82, in fit
    X = super().fit(X)
  File "D:\feature_engine\feature_engine\base_transformers.py", line 98, in fit
    self.variables_ = _find_or_check_numerical_variables(X, self.variables)
  File "D:\feature_engine\feature_engine\variable_manipulation.py", line 82, in _find_or_check_numerical_variables
    raise TypeError(
TypeError: Some of the variables are not numerical. Please cast them as numerical before using this transformer
```
Here instead variables is passed as a str (which _check_input_parameter_variables __allows__) but _find_or_check_numerical_variables doesn't actually run any checks, so the logic goes on up until the line where the column is checked for the presence of inf, which fails because City is an object variable, and should not have made it that far in the code
```
>>> pt = PowerTransformer(variables="City")   
>>> pt.fit(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\feature_engine\feature_engine\transformation\power.py", line 82, in fit
    X = super().fit(X)
  File "D:\feature_engine\feature_engine\base_transformers.py", line 102, in fit
    _check_contains_inf(X, self.variables_)
  File "D:\feature_engine\feature_engine\dataframe_checks.py", line 127, in _check_contains_inf
    if np.isinf(X[variables]).values.any():
  File "D:\Vari\Anaconda3\envs\ML\lib\site-packages\pandas\core\generic.py", line 1936, in __array_ufunc__
    return arraylike.array_ufunc(self, ufunc, method, *inputs, **kwargs)
  File "D:\Vari\Anaconda3\envs\ML\lib\site-packages\pandas\core\arraylike.py", line 366, in array_ufunc
    result = mgr.apply(getattr(ufunc, method))
  File "D:\Vari\Anaconda3\envs\ML\lib\site-packages\pandas\core\internals\managers.py", line 425, in apply
    applied = b.apply(f, **kwargs)
  File "D:\Vari\Anaconda3\envs\ML\lib\site-packages\pandas\core\internals\blocks.py", line 378, in apply
    result = func(self.values, **kwargs)
TypeError: ufunc 'isinf' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
>>>
```
Sorry If I insist of this one but I really need to know what to do to proceed because this issue clearly applies to _find_or_check_datetime_variables aswell

Reply:
Well spotted. It is indeed a bug in both the `_find_or_check_numerical_variables` and the _find_or_check_caterical_variables`. 

If the user passes a string instead of a list, we are not currently checking that the variable is of the corresponding type. 

I create a separate issue to address this bug: